### PR TITLE
WIP: add customisation with parameters to workflows email's body 

### PR DESCRIPTION
**Pull request by Gbaguidi**
Hello Sukhpal Saini. Hope you're doing well

At our Company Gurulab(https://www.linkedin.com/company/gurulab-sas) we are using your work on candymail for email automation. 

But we realise that **it is impossible to customise email's body for each recipient** .

Througth this pull request we add a function **setBodyParameters**, and modify the **runWorkflow** and **buildEmailAction** function to match the setBodyParameters function. 

The function take 2 parameters : **body** and **params**. Params is an array of type  BodyParam = {  key: string,value: string}. 

The keyword to search within the body for replacement is : **PARAMS_key**. 

Each match between the different BodyParam and keyword within the body is replaced by the corresponding value of BodyParam.

Please take a look at the work and give use your feedback.

### CODE REVIEW

File: src/workflow.ts [hunk 1]
```diff
@@ -12,13 +12,16 @@ const getWorkflows = (): Workflow[] => {
...
   if (!getWorkflows() || isEmpty(getWorkflows())) {
     throw new Error('No workflows found. Run the init first: init()')
   }
   const messagesInWorkflow = getWorkflows().find((message) => message.name === workflow)
   if (messagesInWorkflow) {
-    await buildEmailAction(messagesInWorkflow.emails, sendTo)
+    if (typeof params !== "undefined") {
+      await buildEmailAction(messagesInWorkflow.emails, sendTo, params);
+    }
+    await buildEmailAction(messagesInWorkflow.emails, sendTo);
```
Comment (bdcorps) [inline src/workflow.ts:24]: Are we missing an `else` here?

File: src/automation.ts [hunk 2]
```diff
@@ -19,15 +19,26 @@ const init = async (workflows: Workflow[], options: Options) => {
   setConfig(options)
 }
 
-const buildEmailAction = (emails: EmailAction[], sendTo: string) => {
+const buildEmailAction = (emails: EmailAction[], sendTo: string, params?:BodyParam[]) => {
```
Comment (bdcorps) [inline src/automation.ts:22]: Let's make this strict by removing the optional and setting a default value `params: BodyParam[] = []`

File: src/automation.ts [hunk 3]
```diff
@@ -19,15 +19,26 @@ const init = async (workflows: Workflow[], options: Options) => {
...
   emails.forEach(async ({ sendDelay, subject, body, from }) => {
     const template = 'default'
     const today = moment.utc()
     const sendAt = today.add(sendDelay, 'hours').toDate()
-
+    if (typeof params !== "undefined") {
+      body=setBodyParameters(body,params)
```
Comment (bdcorps) [inline src/automation.ts:28]: We should not change the incoming parameter directly. Instead, we can define a new variable like so:
```
let formattedBody = body;
formattedBody = setBodyParameters(body,params);
```

File: examples/simple/index.js [hunk 4]
```diff
@@ -37,6 +37,7 @@ candymail
   })
 
 const someConditionSatisfiedByUser = async () => {
+  params=[{key:"Firstname",value:"Harry"},{key:"Lastname",value:"GBAGUIDI"}]
   const user = process.env.RECIPIENT_EMAIL
-  candymail.runWorkflow('workflow1', user)
+  candymail.runWorkflow('automation1', user,params)
 }
```
Comment (bdcorps) [inline examples/simple/index.js:43]: Awesome job updating the example!

File: __tests__/body-customisation.test.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+  test('Test setBodyParameters with 2 parameters set', async () => {
+    const body = "<h1>PARAMS_Lastname PARAMS_Firstname Send automated messages with Candymail</h1><p>Now with HTML support</p><a href='https://saasbase.dev/candymail'>Learn more here</a>"
+    const params=[{key:"Firstname",value:"Harry"},{key:"Lastname",value:"GBAGUIDI"}]
+    let response = setBodyParameters(body, params)
+    let counter=0
+    params?.forEach(element => {
+      counter=counter+(response.match(/PARAMS_/g) || []).length
+    });
+    expect(counter).toBe(0)
```
Comment (bdcorps) [inline __tests__/body-customisation.test.ts:16]: We could prob shorten this to: `counter = (response.match(/PARAMS_/g) || []).length;` to see if there is any PARAMS_ that wasn't replaced.

File: __tests__/body-customisation.test.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { setBodyParameters} from '../src/automation'
+import { BodyParam} from '../src/types/index'
+import { sampleMessages } from '../src/utils/setupTests'
+import typeorm = require('typeorm')
+
+describe('Unit Tests', () => {
+
+  test('Test setBodyParameters with 2 parameters set', async () => {
+    const body = "<h1>PARAMS_Lastname PARAMS_Firstname Send automated messages with Candymail</h1><p>Now with HTML support</p><a href='https://saasbase.dev/candymail'>Learn more here</a>"
```
Comment (bdcorps) [inline __tests__/body-customisation.test.ts:9]: Fantastic job writing these tests 🤩

File: __tests__/body-customisation.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    let response = setBodyParameters(body, params)
+    let counter=0
+    params?.forEach(element => {
+      counter=counter+(response.match(/PARAMS_/g) || []).length
+    });
+    expect(counter).toBe(0)
+  })
+
+  test('Test setBodyParameters without passing parameters', async () => {
+    const body = "<h1>PARAMS_Lastname PARAMS_Firstname Send automated messages with Candymail</h1><p>Now with HTML support</p><a href='https://saasbase.dev/candymail'>Learn more here</a>"
```
Comment (bdcorps) [inline __tests__/body-customisation.test.ts:20]: What do you think about: using `{}` vs. `PARAMS_`. The former option is used by most email automation software to represent replaceable values. 

For example - `{firstname}`

**Review by bdcorps — CHANGES_REQUESTED** [review]
I'm absolutely thrilled with this PR and your effort. Thank you so much for putting this together. This is a great feature and I would love to include it in the main product.

I've left some comments, mostly around best practices. Let me know if something needs more explanation. Excited to see this go through :D

File: src/workflow.ts (hunk 1 above)
Comment (Gbaguidi) [reply src/workflow.ts:24]: Yes that's true

File: src/automation.ts (hunk 2 above)
Comment (Gbaguidi) [reply src/automation.ts:22]: Done 👍

File: src/automation.ts (hunk 3 above)
Comment (Gbaguidi) [reply src/automation.ts:28]: I've change it to this :
`
let formattedBody = body;
    if (typeof params !== "undefined") {
      formattedBody=setBodyParameters(body,params)
    }
    await addScheduledMessage({ template, sendFrom, sendTo, sendAt, subject, body:formattedBody })
`

File: examples/simple/index.js (hunk 4 above)
Comment (Gbaguidi) [reply examples/simple/index.js:43]: Thank you :)

File: __tests__/body-customisation.test.ts (hunk 6 above)
Comment (Gbaguidi) [reply __tests__/body-customisation.test.ts:9]: My pleasure :)

File: __tests__/body-customisation.test.ts (hunk 7 above)
Comment (Gbaguidi) [reply __tests__/body-customisation.test.ts:20]: - Yes at first i've thought the same way but, for email with code snipets and thing like that the end user of the package could probably write somethings like object `car={color:"red", brand:"ferrari", ...}` which could trigger errors at some point (i think). I don't know :sweat_smile: .

Comment (bdcorps) [reply __tests__/body-customisation.test.ts:15]: That's definitely an extended use case - to be able to pass in object (the possibilities!). For this PR, let's assume it will only a string.

Comment (bdcorps) [reply __tests__/body-customisation.test.ts:15]: Actually let's make it double curly braces - `{{firstname}}`

**Review by bdcorps — CHANGES_REQUESTED** [review]
Coming along well! The last remaining item is just reformatting how the placeholder looks to the end user. We should be good to merge once it's implemented.

File: __tests__/body-customisation.test.ts (hunk 7 above)
Comment (Gbaguidi) [reply __tests__/body-customisation.test.ts]: Okey, i've made the change. It should be good now. Let's ship it 😍.
